### ISSUE TITLE
A more flexible target definition for relations

### ISSUE THREAD

**Issue by breitmark** [adam-cowley/neode#18]
Within Neo4j it is possible to use a single request with different labelled nodes like 
MATCH (u)-[works_on]->(w) WHERE u.name="John" RETURN w
With a user u named John, w could be a company, department, project.

In Neode it should then be something like:

`module.exports = {
    name: {
        type: 'string',
        indexed: true,
        required: true
    },

    work: {
        type: 'relationship',
        relationship: 'work',
        direction: 'out',
        target: ['Company', 'Department', 'Workgroup', 'Project'],
    }
};`
 or even target: [] for any kind of target

**Comment by adam-cowley** [comment]
This sounds like a fair request but does throw up some potential issues.  At the moment neode will attempt to automatically create the nested/eager nodes from the supplied object. Eg. if you provide an object for `work`, it will try to create the target too.

```
{
  name: 'Adam',
  work: {
    name: 'Neo4j',
  },
}
```

If you provide an array of targets, when it comes to creating/merging a record it will either need to:
a) Throw an Error - "Cannot create a relationship with multiple targets"
b) Require that creating/merging passes a Node instance for multiple targets
c) Take the first label and attempt to create with that
d) Somehow attempt to work out from the object what has been passed through.

In the case of a and b, you could always set up explicit relationships for creating relationships, eg:

```
work_Company: {
    type: 'relationship',
    relationship: 'work',
    direction: 'out',
    target: 'Company',
}
```

I'd love to hear your feedback.  How do you see this being used in your application?

While we're at it, I could add support for an array of relationships for eager loading. This is something that I can see being useful in a real world scenario (a Person likes a Post, Comment or Page) but would also support multiple specific types (`-[:LIKES_POST]->`, `-[:LIKES_COMMENT]->` etc).

Again, this could throw an error if you try to create or take the first type and try to validate against that.

```
relationship: ['LIKES_POST', 'LIKES_COMMENT']
```

**Comment by breitmark** [comment]
My starting thought is what would happen with:
Neode.cypher('MATCH (u)-[works_on]->(w) WHERE u.name="John" RETURN w').then(function(result) {
   const targets = Neode.hydrate(result, 'w');
}

For me, the method (c) would be fine as I currently only use single labelled nodes. For more sophisticated use cases you could add the possibility to add callbacks to the model.

**Comment by adam-cowley** [comment]
With your example, it will take each node and hydrate it with the relevant model information as long as a definition matches the labels of the node.   

As an unintended side effect, if you provide a node/relationship and don't define the target then all nodes at the end of that relationship will be returned.  For example:

```
    relationshipToAnything: {
            type: 'relationship',
            relationship: 'RELATIONSHIP',
            direction: 'out',
            properties: {},
            eager: true,
            alias: 'node',
            properties: {
                since: {
                    type: 'int',
                    default: Date.now
                }
            }, 
        },
        nodeToAnything: {
            type: 'node',
            relationship: 'RELATIONSHIP',

            direction: 'out',
            eager: true,
        },
```

I've written the logic in to throw an error if no definition is defined when you try to create/merge a node.  It'll be available and tested in the @next/0.2.5 release.

**Comment by breitmark** [comment]
I was aware of that "unintended side effect" and have a use case for it. With (u:User)-[w:worksOn]->(p) the type of p determines which kind of editor to open for the users work
